Adds course block location to the ProblemResponse report, sourced from REST API

pomegranited opened the pull request:
This change provides "course location" information and sorting for problems included in the Problem Response report.

* Course and Course Blocks data are stored to hive tables, partitioned by a formatted datetime string.
* Problem Response records are also partitioned by a formatted datetime string.
* Adds the EdxRestApiTask for contacting the REST API and caching responses.  This adds a dependency on edx-rest-api-client and its dependencies, but these modules are only required on the pipeline host machine, not the EMR instances.

**Discussions**: This change is related to [edx-analytics-data-api PR #128](https://github.com/edx/edx-analytics-data-api/pull/128), which has been discussed extensively with the Analytics team.

**Dependencies**: This is an internal pull request, intended to be merged with the [upstream PR #274](https://github.com/edx/edx-analytics-pipeline/pull/274).

**Partner information**: 3rd party-hosted open edX instance

**Testing instructions**:

1. See [edx-analytics-data-api PR 128](https://github.com/edx/edx-analytics-data-api/pull/128) for analyticstack test setup instructions.  See below for notes and alterations.
2. Before running `remote-task` to install the analytics-pipeline, create an OAuth2 client:

    ```
sudo su edxapp
./manage.py lms --settings=devstack create_oauth2_client  \
            http://localhost:9999 \
            http://localhost:9999/complete/edx-oidc/  \
            confidential \
            --client_name "Analytics Pipeline" \
            --client_id oauth_id \
            --client_secret oauth_secret \
            --trusted
            --user=staff
```
3. Create a new `override.cfg` file by copying `config/devstack.cfg`. Update the parameters under `[edx-rest-api]` with your chosen OAuth2 `client_id` and `client_secret`.
    * Note that variables in `override.cfg` should not be quoted.  The quotes are taken literally, not interpreted.
4. You don't need to run the Analytics API setup, or the custom ElasticSearch.
5. You *do* need to run the edxapp LMS service, which provides the REST API used to source the course and course block data.  You can check that the edX REST APIs are functioning by visiting:
  * [http://localhost:8000/api/courses/v1/courses/](http://localhost:8000/api/courses/v1/courses/)
  * [http://localhost:8000/api/courses/v1/blocks/](http://localhost:8000/api/courses/v1/blocks/?course_id=course-v1:edX%2BDemoX%2BDemo_Course&all_blocks=true&depth=all).  Select a course, and append `all_blocks=true&depth=all&requested_fields=children` to your querystring to see the full block hierarchy.  
6. Run the `LoadAllCourseBlocksWorkflow` task to install the pipeline from this PR's branch, and load all the courses and course blocks into your hive tables:

    ```
export PIPELINE_REPO="https://github.com/open-craft/edx-analytics-pipeline.git"
export PIPELINE_BRANCH="jill/block-location"
export WHEEL_URL=http://edx-wheelhouse.s3-website-us-east-1.amazonaws.com/Ubuntu/precise
export VAGRANT_PATH='..'  # or wherever your Vagrantfile is
export REMOTE_NAME='analytics-block-location'
remote-task --wheel-url $WHEEL_URL --vagrant-path "$VAGRANT_PATH" \
    --repo "$PIPELINE_REPO" --branch "$PIPELINE_BRANCH" \
    --remote-name "$REMOTE_NAME"  --override-config `pwd`/override.cfg  --wait \
    LoadAllCourseBlocksWorkflow --local-scheduler --n-reduce-tasks 2 --remote-log-level DEBUG
```
1. To view the course data, query the hive table created by the task:

    ```
sudo su hadoop
hive
hive> select * from course_list;
hive> select * from course_blocks;
```
1. To see the course location field appear in the Problem Response report, run this task on the installed pipeline:

   ```
sudo su hadoop
cd /var/lib/analyticstack/analytics-block-location
source venv/bin/activate
cd repo
launch-task ProblemResponseReportWorkflow --local-scheduler --n-reduce-tasks 2 --remote-log-level DEBUG --marker hdfs://localhost:9000/reports/marker`date +%s`
```
1. Check the configured `report_output_root` to see a report generated for each course in your LMS that has had problems answered.  The devstack.cfg file stores them under `/edx/src/problem-response-reports/`.
1. Run unit tests:

    ```
vagrant ssh
cd /var/lib/analytics-tasks/analyticstack/repo
source ../venv/bin/activate
export WHEEL_URL=http://edx-wheelhouse.s3-website-us-east-1.amazonaws.com/Ubuntu/precise
export WHEEL_PYVER=python2.7
make coverage
```

**Author notes and concerns**:

1. I still need to add acceptance tests.
1. Will merge this to the upstream PR branch for edX to review.

**Reviewers**
- [ ] @haikuginger

haikuginger commented on config/devstack.cfg:
```diff
@@ -86,4 +86,23 @@ alias = roster
 
 [problem-response]
-report_fields = ["username","problem_id","answer_id","problem","question","score","max_score","correct","answer","total_attempts","first_attempt_date","last_attempt_date"]
+report_fields = ["username","problem_id","answer_id","location","question","score","max_score","correct","answer","total_attempts","first_attempt_date","last_attempt_date"]
```
Might be easier to visually diff this field later on if you broke each item out onto its own line.

pomegranited replied:
Addressed with [6caeade](https://github.com/open-craft/edx-analytics-pipeline/pull/4/commits/6caeade8d9dc7dd8e36cab743418486af8a1481a).

haikuginger commented on edx/analytics/tasks/course_blocks.py:
```diff
@@ -0,0 +1,501 @@
+    parent_block_id = StringField(length=255, nullable=True,
+                                  description='Block identifier for the block\'s parent. If null, and '
+                                              'is_root and has_multiple_parents are False, then this block has been '
+                                              'removed from the course.')
```
If we can derive this information, why not include it as a field?

pomegranited replied:
Good point.  I added `is_orphan`, and so that made me want to rename `has_multiple_parents` to `is_dac`, which looks a little more cryptic, but makes the boolean field names nice and consistent.

See [99ee83b](https://github.com/open-craft/edx-analytics-pipeline/pull/4/commits/99ee83b33f7b100b3bfb5aab83a67d606cad27d9).

haikuginger commented on edx/analytics/tasks/course_blocks.py:
```diff
@@ -0,0 +1,501 @@
+    partition_format = luigi.Parameter(
+        config_path={'section': 'course_blocks', 'name': 'partition_format'},
+        default='%Y%m%d',
+        description="Format string for the course blocks table partition's `date` parameter.n"
```
Nit: Consistently using the same quote style would be nice - to switch to single-quotes here would require some escaping, though, so it's not strictly necessary.

pomegranited replied:
Addressed with [b535690](https://github.com/open-craft/edx-analytics-pipeline/pull/4/commits/b53569043af8fd18975cf9a31d70bf935c9f7fa3).

haikuginger commented on edx/analytics/tasks/course_blocks.py:
```diff
@@ -0,0 +1,501 @@
+      also have `parent_block_id=None`, `course_path=''`, and `sort_idx=0`.
+    * child: Blocks which have a single parent will have `parent_block_id not None`, and a `course_path` string which
+      concatenates the parent block's display_name values.
+    * DAC: Blocks which have more than one parent (and thus create directed acyclic graphs, aka DACs) are marked with
```
DAC or DAG?

pomegranited replied:
Addressed with [99ee83b](https://github.com/open-craft/edx-analytics-pipeline/pull/4/commits/99ee83b33f7b100b3bfb5aab83a67d606cad27d9).

haikuginger commented on edx/analytics/tasks/course_blocks.py:
```diff
@@ -0,0 +1,501 @@
+    deleted_blocks_path = luigi.Parameter(
+        config_path={'section': 'course-blocks', 'name': 'deleted_blocks_path'},
+        default='(Deleted block :)',
+        description='Mark deleted (unparented) blocks with this string in course_path.',
```
Use the "orphan" terminology from higher up?

pomegranited replied:
Addressed with [99ee83b](https://github.com/open-craft/edx-analytics-pipeline/pull/4/commits/99ee83b33f7b100b3bfb5aab83a67d606cad27d9).

haikuginger commented on edx/analytics/tasks/course_blocks.py:
```diff
@@ -0,0 +1,501 @@
+            more = []
+
+        children = root.get('children', [])
+        more += children
```
This will result in us visiting the same node multiple times if it has multiple parents (and visit all its children again). Possibly add a `visited` set to track which nodes we've been to?

pomegranited replied:
Addressed with [3e76559](https://github.com/open-craft/edx-analytics-pipeline/pull/4/commits/3e7655964bf18763e05462fa4301134b99e3d974).

haikuginger commented on edx/analytics/tasks/course_blocks.py:
```diff
@@ -0,0 +1,501 @@
+
+        https://github.com/spotify/luigi/blob/master/doc/tasks.rst#dynamic-dependencies
+
+    So we have to trick luigi into allowing us to have dynamic dependencies by changing the requirements during the
```
This is fine.

haikuginger commented on edx/analytics/tasks/course_list.py:
```diff
@@ -0,0 +1,253 @@
+
+        Yields the CourseRecord as a tuple.
+        """
+        # Note that there should only ever by one record in the values list,
```
s/by/be/

pomegranited replied:
Addressed with [b535690](https://github.com/open-craft/edx-analytics-pipeline/pull/4/commits/b53569043af8fd18975cf9a31d70bf935c9f7fa3).

haikuginger commented:
I've manually tested the code, and it behaves as I expect it to, so once we've addressed the comments above, a thumbs-up should be forthcoming.

pomegranited commented:
@haikuginger I've addressed your comments, and also included [101cb88](https://github.com/open-craft/edx-analytics-pipeline/pull/4/commits/101cb8889a0eb8013e6617d053ac6988637eebe2), which addresses 2 major issues with the LoadAllCourseBlocksWorkflow and ProblemResponseReportWorkflow tasks discovered in production.
1. The first issue was due to the tracking.log naming scheme used in production, and how logs are filtered for a given date interval. 
2. The second issue occurred in production but not dev, because of the ephemeral EMR instances used.  These instances contain an empty hive instance, which must be populated with tables and partitions from data that has been stored to S3.

To test these fixes:
1. Make a datestamped copy of your existing tracking.log file:
   
   ```
   sudo su hadoop
   hdfs dfs -cp /data/tracking.log /data/tracking.log.20160828
   ```
2. Clear out your hive tables:
   
   ```
       sudo su hadoop
       hive
       hive> show tables;
       hive> drop table course_blocks;
       hive> drop table course_list;
       hive> drop table problem_response_latest;
       hive> drop table problem_response_location;
   ```
3. Run the `LoadAllCourseBlocksWorkflow` task.
4. Go back into hive, and drop the `course_blocks` and `course_list` tables again.
5. Run the `ProblemResponseReportWorkflow` task.
6. Check that the report output is still generated as expected, and contains the course blocks location data.
   Doubling up the tracking logs should not affect the report output, except for the `ProblemResponseRecord.total_attempts` field, which, if it is present in your report, will be doubled.

Could you please re-review, and let me know if you need any changes made?

haikuginger commented on config/devstack.cfg:
```diff
@@ -86,4 +86,34 @@ alias = roster
 
 [problem-response]
-report_fields = ["username","problem_id","answer_id","problem","question","score","max_score","correct","answer","total_attempts","first_attempt_date","last_attempt_date"]
+report_fields = ["username",
```
Nit: Move to the standard multiline spacing, like so:

``` python
[
    "item1",
    "item2",
    "item3",
]
```

haikuginger commented:
:+1:
- [x] I tested this: I generated data on a local analyticstack and ran the reports provided; the report contained course block location information as expected.
- [x] I read through the code
- [x] Includes documentation in the form of code comments/docstrings

pomegranited commented:
Rebased from [c95cc13](https://github.com/edx/edx-analytics-pipeline/pull/274/commits/c95cc13d9800cb880850e82de8a81ca527bee333) to ease upstream review.
